File: Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        pmap = {"{" : "}", "(" : ")", "[" : "]"}
        pre = {"{", "(", "["}
        
        stack = []
        
        for i in s:
            if i in pre:
                stack.append(i)
            else:
                if not stack:
                    return False
                    
                elif pmap[stack.pop()] != i:
                    return False
        if stack:
            return False
        return True

File: Valid_Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if (s.empty()) return true;
        stack<char> ps;
        unordered_set<char> set = {'(', '{', '['};
        unordered_map<char, char> map = {{'(', ')'}, {'{', '}'}, {'[', ']'}};
        for (string::iterator it = s.begin(); it != s.end() ; ++it){
            if (set.count(*it)) ps.push(*it);
            else{
                if (ps.empty()) return false;
                if (map[ps.top()] == *it){
                    ps.pop();
                }
                else return false;
            }
        }
        
        return ps.empty();
    }
};